### PR TITLE
Add Google Batch NOT_FOUND error management

### LINKED ISSUE #5422: NOT_FOUND error on google-batch
## Bug report 

I have a RNA-seq pipeline I've been using for a while on GCP with success. Recently, I noticed that when processing medium/large datasets, I get a `NOT_FOUND` error and the workflow gets interrupted. It feels like a temporary unavailable service, but it keeps reoccurring when resuming the run.

### Expected behavior and actual behavior

No interruption or retry if a given service is not available

### Steps to reproduce the problem

Hard to reproduce, could be due to temporarily unavailable services

### Program output 

```
Oct-22 16:33:28.660 [Task monitor] DEBUG nextflow.processor.TaskProcessor - Handling unexpected condition for
  task: name=ROME_QUANT_HUMAN:RNASEQC (ERR10232708); work-dir=gs://nxf-work/romequant/cedric/c9/cec2987387dd8a25f738dab2db92a9
  error [com.google.api.gax.rpc.NotFoundException]: io.grpc.StatusRuntimeException: NOT_FOUND: Resource "projects/114884310888/locations/us-central1/jobs/nf-c9cec298-1729608617842/taskGroups/group0/tasks/0" was not found
Oct-22 16:33:28.683 [Task monitor] ERROR nextflow.processor.TaskProcessor - Error executing process > 'ROME_QUANT_HUMAN:RNASEQC (ERR10232708)'

Caused by:
  NOT_FOUND: Resource "projects/114884310888/locations/us-central1/jobs/nf-c9cec298-1729608617842/taskGroups/group0/tasks/0" was not found

com.google.api.gax.rpc.NotFoundException: io.grpc.StatusRuntimeException: NOT_FOUND: Resource "projects/114884310888/locations/us-central1/jobs/nf-c9cec298-1729608617842/taskGroups/group0/tasks/0" was not found
        at com.google.api.gax.rpc.ApiExceptionFactory.createException(ApiExceptionFactory.java:90)
        at com.google.api.gax.grpc.GrpcApiExceptionFactory.create(GrpcApiExceptionFactory.java:98)
        at com.google.api.gax.grpc.GrpcApiExceptionFactory.create(GrpcApiExceptionFactory.java:66)
        at com.google.api.gax.grpc.GrpcExceptionCallable$ExceptionTransformingFuture.onFailure(GrpcExceptionCallable.java:97)
        at com.google.api.core.ApiFutures$1.onFailure(ApiFutures.java:84)
        at com.google.common.util.concurrent.Futures$CallbackListener.run(Futures.java:1127)
        at com.google.common.util.concurrent.DirectExecutor.execute(DirectExecutor.java:31)
        at com.google.common.util.concurrent.AbstractFuture.executeListener(AbstractFuture.java:1286)
        at com.google.common.util.concurrent.AbstractFuture.complete(AbstractFuture.java:1055)
        at com.google.common.util.concurrent.AbstractFuture.setException(AbstractFuture.java:807)
        at io.grpc.stub.ClientCalls$GrpcFuture.setException(ClientCalls.java:568)
        at io.grpc.stub.ClientCalls$UnaryStreamToFuture.onClose(ClientCalls.java:538)
        at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
        at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
        at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
        at com.google.api.gax.grpc.ChannelPool$ReleasingClientCall$1.onClose(ChannelPool.java:570)
        at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:567)
        at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:71)
        at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:735)
        at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:716)
        at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
        at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:133)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at java.base/java.lang.Thread.run(Thread.java:840)
        Suppressed: com.google.api.gax.rpc.AsyncTaskException: Asynchronous task failed
                at com.google.api.gax.rpc.ApiExceptions.callAndTranslateApiException(ApiExceptions.java:57)
                at com.google.api.gax.rpc.UnaryCallable.call(UnaryCallable.java:112)
                at com.google.cloud.batch.v1.BatchServiceClient.getTask(BatchServiceClient.java:789)
                at com.google.cloud.batch.v1.BatchServiceClient.getTask(BatchServiceClient.java:732)
                at nextflow.cloud.google.batch.client.BatchClient$_describeTask_lambda4.doCall(BatchClient.groovy:113)
                at dev.failsafe.Functions.lambda$toCtxSupplier$11(Functions.java:236)
                at dev.failsafe.Functions.lambda$get$0(Functions.java:46)
                at dev.failsafe.internal.RetryPolicyExecutor.lambda$apply$0(RetryPolicyExecutor.java:75)
                at dev.failsafe.SyncExecutionImpl.executeSync(SyncExecutionImpl.java:176)
                at dev.failsafe.FailsafeExecutor.call(FailsafeExecutor.java:437)
                at dev.failsafe.FailsafeExecutor.get(FailsafeExecutor.java:115)
                at nextflow.cloud.google.batch.client.BatchClient.apply(BatchClient.groovy:187)
                at nextflow.cloud.google.batch.client.BatchClient.describeTask(BatchClient.groovy:113)
                at nextflow.cloud.google.batch.client.BatchClient.getTaskStatus(BatchClient.groovy:122)
                at nextflow.cloud.google.batch.GoogleBatchTaskHandler.getTaskState(GoogleBatchTaskHandler.groovy:454)
                at nextflow.cloud.google.batch.GoogleBatchTaskHandler.checkIfRunning(GoogleBatchTaskHandler.groovy:479)
                at nextflow.processor.TaskPollingMonitor.checkTaskStatus(TaskPollingMonitor.groovy:642)
                at nextflow.processor.TaskPollingMonitor.checkAllTasks(TaskPollingMonitor.groovy:571)
                at nextflow.processor.TaskPollingMonitor.pollLoop(TaskPollingMonitor.groovy:441)
                at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
                at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
                at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
                at java.base/java.lang.reflect.Method.invoke(Method.java:568)
                at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:343)
                at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:328)
                at groovy.lang.MetaClassImpl.doInvokeMethod(MetaClassImpl.java:1333)
                at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1088)
                at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1007)
                at org.codehaus.groovy.runtime.InvokerHelper.invokePogoMethod(InvokerHelper.java:645)
                at org.codehaus.groovy.runtime.InvokerHelper.invokeMethod(InvokerHelper.java:628)
                at org.codehaus.groovy.runtime.InvokerHelper.invokeMethodSafe(InvokerHelper.java:82)
                at nextflow.processor.TaskPollingMonitor$_start_closure2.doCall(TaskPollingMonitor.groovy:316)
                at nextflow.processor.TaskPollingMonitor$_start_closure2.call(TaskPollingMonitor.groovy)
                at groovy.lang.Closure.run(Closure.java:505)
                ... 1 common frames omitted
Caused by: io.grpc.StatusRuntimeException: NOT_FOUND: Resource "projects/114884310888/locations/us-central1/jobs/nf-c9cec298-1729608617842/taskGroups/group0/tasks/0" was not found
        at io.grpc.Status.asRuntimeException(Status.java:537)
        ... 14 common frames omitted
Oct-22 16:33:28.698 [Task monitor] DEBUG nextflow.Session - Session aborted -- Cause: io.grpc.StatusRuntimeException: NOT_FOUND: Resource "projects/114884310888/locations/us-central1/jobs/nf-c9cec298-1729608617842/taskGroups/group0/tasks/0" was>
Oct-22 16:33:28.750 [Task monitor] DEBUG nextflow.Session - The following nodes are still active:
```

### Environment 

* Nextflow version: 24.09.2-edge build 5927
* Java version: openjdk version "11.0.24" 2024-07-16
* Operating system: Ubuntu
* Bash version: zsh 5.8

### LINKED ISSUE #5727: NullPointerException when retrying task with an input transfer failure
## Bug report 

A NullPointerException is thrown during a task retry when a failure there is a failure in the foreign file staging of an input file managed by the file porter. (https://github.com/nextflow-io/nextflow/pull/5690#issuecomment-2614781547)

### Expected behavior and actual behavior

When a task requires an input file transfer and it fails, Nextflow produces an exception during the `invokeTask` method before generating the task hash, the management of the exception creates a retry but, it fails because the hash is null.  

### Steps to reproduce the problem

The original failure happens when a read timeout exception is produced when transferring a file in an s3 bucket to a working dir in Google Cloud. I have reproduced the same error when using a non-existing file in s3 as task input, and running the pipeline in Google Cloud. When running the same locally the task fails without retrying.

### Program output 

[See attached log](https://github.com/user-attachments/files/18553111/default.nextflow.log).

### Environment 

* Nextflow version: 24.12.0-edge 
* Java version: 21
* Operating system: Linux

### PR DESCRIPTION
**Pull request by jorgee**
close #5422
close #5727

This PR includes a possible fix for the NotFoundException returned by the Google Batch API when getting some tasks.

When the client.getTaskStatus throws a NotFoundException it is caught and managed in the following way:
- If the retrieved from client.listTasks has several tasks. It tries to find the task and check the status 
- otherwise or if the task is not found in the list it retrieves the job status.
- If it is not able to neither get the task or job status it returns pending

Including a unit test producing the NotFoundException to validate the logic.

Two corner cases could not be correctly managed. 
1. When neither task status nor a job status is found, it is returning PENDING. I think that it could happen in the initial stage when Google Batch is creating the job and tasks. I am assuming that a task or job status will be received at some point in the execution. It is the same management that we did when no tasks were found in the job. 
2. When a task is not found in Google Batch, it belongs to a task array job and the job status is RUNNING or FAILED, we could get an incorrect task status. It is not very important in the case of RUNNING, but with FAILED we could get a task failed but, its state is unknown. According to the API documentation, the RUNNING or FAILED job states mean that at least one of the tasks is in this state. Job could be FAILED when there has been a job failure (such as the invalid type one we saw in #5550). So, I am assuming that when the task status is not found, it is not in the list and the Job is FAILED, It is due to a job failure and the task is also FAILED. Other alternatives that I have considered are: set the task to PENDING, but I think we could get a deadlock if it is job failure; or throw an exception to abort the execution. Other suggestions are welcome.

### CODE REVIEW

**Comment by pditommaso** [comment]
Should @ejseqera run stress test on this?

**Comment by ejseqera** [comment]
I'm on it

**Comment by ewels** [comment]
@jorgee I know we just discussed on it on the call, but can we confirm that this PR also handles these states please?

* `UNAVAILABLE`
* `DEADLINE_EXCEEDED`
* `RESOURCE_EXHARSTED`
* `UNKNOWN`
* ..more

Thanks!

**Comment by jorgee** [comment]
> @jorgee I know we just discussed on it on the call, but can we confirm that this PR also handles these states please?
> 
> * `UNAVAILABLE`
> * `DEADLINE_EXCEEDED`
> * `RESOURCE_EXHARSTED`
> * `UNKNOWN`
> * ..more
> 
> Thanks!

I can't see these states in the Google Batch API

**Comment by robnewman** [comment]
@jorgee https://github.com/nextflow-io/nextflow/issues/4537

**Comment by pditommaso** [comment]
https://github.com/nextflow-io/nextflow/issues/4537 is a different issue

**Comment by jorgee** [comment]
Ok, you mean the code in the exception. No, I did it is only for NOT_FOUND, because it was not solved even with long retries.

**Comment by jorgee** [comment]
> @jorgee We're trying to determine if this PR solves [this customer feature request](https://feedback.seqera.io/feature-requests/p/nextflow-to-handle-google-grpc-api-call-failure)

No, this PR is not solving it. You can find the current retriable exceptions [here](https://github.com/nextflow-io/nextflow/blob/2153abb1491fa765f56f7e3fcab0cce66b14c011/plugins/nf-google/src/main/nextflow/cloud/google/batch/client/BatchClient.groovy#L176l)

**Comment by robnewman** [comment]
@jorgee New issue reported: [Nextflow to gracefully handle Google gRPC API call failures](https://github.com/nextflow-io/nextflow/issues/5703)

**Comment by pditommaso** [comment]
@ejseqera any update on this? let us know if you need assistance for this branch

File: plugins/nf-google/src/main/nextflow/cloud/google/batch/GoogleBatchTaskHandler.groovy [hunk 1]
```diff
@@ -459,23 +462,45 @@ class GoogleBatchTaskHandler extends TaskHandler implements FusionAwareTask {
...
+
+    protected String manageNotFound( Iterable<Task> tasks) {
+        // If task is array, check if the in the task list
+        if (tasks.size() > 1) {
+            for (Task t in tasks) {
+                if (t.name == client.generateTaskName(jobId, taskId)) {
+                    inspectTaskStatus(t.status)
+                    return taskState
+                }
+            }
```
Comment (pditommaso) [inline plugins/nf-google/src/main/nextflow/cloud/google/batch/GoogleBatchTaskHandler.groovy:498]: not sure to understand how status in the case of a array of tasks is handled?

**Comment by ejseqera** [comment]
I've attempted a large nf-core/sarek stress test run (~1390 jobs) and ran into several igenomes download issues and more importantly, several `java.lang.NullPointerException: Cannot invoke "com.google.common.hash.HashCode.asBytes()" because "hash" is null` in the `TaskProcessor` suggesting there might be an underlying issue with task submission/retry logic. [See attached log](https://github.com/user-attachments/files/18553111/default.nextflow.log).

Will update once I have results from the rerun with local reference data but this seems unrelated?

File: plugins/nf-google/src/main/nextflow/cloud/google/batch/GoogleBatchTaskHandler.groovy (hunk 1 above)
Comment (jorgee) [reply plugins/nf-google/src/main/nextflow/cloud/google/batch/GoogleBatchTaskHandler.groovy:498]: There are two ways to check the task status:  the `getTaskStatus` that is failing or looking for the task in the job's task list retrieved at the beginning of the `getTaskState` method. As the first option fails, this is trying with the second option.

**Comment by jorgee** [comment]
@ejseqera It as an unrelated issue, the stage of the file is failing and in the retry the hash is null. I am looking why the hash is null in this case.

**Comment by bentsherman** [comment]
@jorgee the old implementation of checking job status: #3892

**Comment by jorgee** [comment]
I have updated the branch with the following changes
Add a flag to indicate if a task belongs to array, when a task belongs to an array I use the current way to check the status but with the a fallback to the jobStatus if it fails. For single tasks, it uses the old jobStatus check. 
In this PR #5723, I have includesd an alternative for managing the status of tasks in arrays that reduces the API calls and does not require the specific management of the NotFoundException.

File: plugins/nf-google/src/main/nextflow/cloud/google/batch/GoogleBatchTaskHandler.groovy [hunk 2]
```diff
@@ -91,6 +95,11 @@ class GoogleBatchTaskHandler extends TaskHandler implements FusionAwareTask {
      */
     private String uid
 
+    /**
+     * Flag to indicate if task belong to an TaskRunArray
+     */
+    private boolean belongsToArray
```
Comment (pditommaso) [inline plugins/nf-google/src/main/nextflow/cloud/google/batch/GoogleBatchTaskHandler.groovy:101]: Why not using `Task.isArray` ? 

https://github.com/nextflow-io/nextflow/blob/2a412671df356f801e0ad9a3808e6b99649a362c/modules/nextflow/src/main/groovy/nextflow/processor/TaskRun.groovy#L759-L761

Comment (jorgee) [reply plugins/nf-google/src/main/nextflow/cloud/google/batch/GoogleBatchTaskHandler.groovy:101]: If I am not wrong, it indicates if task is a TaskArrayRun but not if the task belongs to an array

Comment (bentsherman) [reply plugins/nf-google/src/main/nextflow/cloud/google/batch/GoogleBatchTaskHandler.groovy:101]: Only the task array handler is submitted, so it is safe to use `isArray()` during the status check

Comment (jorgee) [reply plugins/nf-google/src/main/nextflow/cloud/google/batch/GoogleBatchTaskHandler.groovy:101]: So, this will change the meaning of isArray. Now, it will mean the task is TaskArrayRun or a TaskRun in an array. If you agree, I will move the boolean flag to TaskRun set it false by default and change when it is collected to a TaskArray.

Comment (pditommaso) [reply plugins/nf-google/src/main/nextflow/cloud/google/batch/GoogleBatchTaskHandler.groovy:101]: yeah, you are right, let's call it `isChild`

Comment (jorgee) [reply plugins/nf-google/src/main/nextflow/cloud/google/batch/GoogleBatchTaskHandler.groovy]: Ok, I have just renamed it

**Comment by jorgee** [comment]
> @ejseqera It as an unrelated issue, the stage of the file is failing and in the retry the hash is null. I am looking why the hash is null in this case.

**Comment by ejseqera** [comment]
@jorgee Should I got ahead and run another test with this updated branch now? Is it worth also testing #5723?

**Comment by jorgee** [comment]
> @jorgee Should I go ahead and run another test with this updated branch now? Is it worth also testing #5723?

Yes, please check with this updated branch. I have also included a change to fix the hash null due to the read timeout. #5723 only affects when using task arrays.

File: plugins/nf-google/src/main/nextflow/cloud/google/batch/GoogleBatchTaskHandler.groovy [hunk 3]
```diff
@@ -91,6 +95,11 @@ class GoogleBatchTaskHandler extends TaskHandler implements FusionAwareTask {
      */
     private String uid
 
+    /**
+     * Flag to indicate if task belong to an TaskRunArray
+     */
+    private boolean isChild
```
Comment (pditommaso) [inline plugins/nf-google/src/main/nextflow/cloud/google/batch/GoogleBatchTaskHandler.groovy:101]: Could keep this as an attribute of `TaskRun` is the same vein `isArray` and set when the object is created?

File: modules/nextflow/src/main/groovy/nextflow/file/FilePorter.groovy [hunk 4]
```diff
@@ -331,7 +331,7 @@ class FilePorter {
                     // remove the target file that could be have partially downloaded
                     cleanup(stagePath)
                     // check if a stage/download retry is allowed
-                    if( count++ < maxRetries && e !instanceof NoSuchFileException && e !instanceof InterruptedIOException && !Thread.currentThread().isInterrupted() ) {
+                    if( count++ < maxRetries && e !instanceof NoSuchFileException && (e instanceof SocketTimeoutException || e !instanceof InterruptedIOException) && !Thread.currentThread().isInterrupted() ) {
```
Comment (pditommaso) [inline modules/nextflow/src/main/groovy/nextflow/file/FilePorter.groovy:334]: Let's move ` e !instanceof NoSuchFileException && (e instanceof SocketTimeoutException || e !instanceof InterruptedIOException)` into a separate method e.g. `recoverableError(e)` otherwise it's impossible to understand it. 

I would also add `SocketException` at this point

Comment (jorgee) [reply modules/nextflow/src/main/groovy/nextflow/file/FilePorter.groovy]: Changed in last commit

File: plugins/nf-google/src/main/nextflow/cloud/google/batch/GoogleBatchTaskHandler.groovy (hunk 3 above)
Comment (jorgee) [reply plugins/nf-google/src/main/nextflow/cloud/google/batch/GoogleBatchTaskHandler.groovy]: Changed in last commit

**Comment by ejseqera** [comment]
While testing this updated branch, I discovered a new issue where the pipeline fails with `UnknownHostException` when trying to access Google Cloud Storage. This appears to happen when the network connection to `storage.googleapis.com` is interrupted or fails. The error occurs in multiple threads when trying to access Google Cloud Storage:
```
Feb-03 09:48:01.090 [Actor Thread 117] ERROR nextflow.extension.OperatorImpl - @unknown
com.google.cloud.storage.StorageException: storage.googleapis.com
	at com.google.cloud.storage.StorageException.translate(StorageException.java:118)
	at com.google.cloud.storage.spi.v1.HttpStorageRpc.translate(HttpStorageRpc.java:287)
	at com.google.cloud.storage.spi.v1.HttpStorageRpc.get(HttpStorageRpc.java:512)
	at com.google.cloud.storage.StorageImpl.lambda$get$6(StorageImpl.java:285)
	at com.google.api.gax.retrying.DirectRetryingExecutor.submit(DirectRetryingExecutor.java:102)
	at com.google.cloud.RetryHelper.run(RetryHelper.java:76)
	at com.google.cloud.RetryHelper.runWithRetries(RetryHelper.java:50)
	at com.google.cloud.storage.Retrying.run(Retrying.java:54)
	at com.google.cloud.storage.StorageImpl.run(StorageImpl.java:1406)
	at com.google.cloud.storage.StorageImpl.get(StorageImpl.java:284)
	at com.google.cloud.storage.StorageImpl.get(StorageImpl.java:290)
	at com.google.cloud.storage.contrib.nio.CloudStorageFileSystemProvider.readAttributes(CloudStorageFileSystemProvider.java:789)
	at java.base/java.nio.file.Files.readAttributes(Files.java:1853)
	at org.codehaus.groovy.vmplugin.v8.IndyInterface.fromCache(IndyInterface.java:321)
	at nextflow.extension.FilesEx.createDirIfNotExists(FilesEx.groovy:1352)
	at nextflow.file.FileCollector.saveTo0(FileCollector.groovy:225)
	at nextflow.file.FileCollector.saveTo(FileCollector.groovy:251)
	at org.codehaus.groovy.vmplugin.v8.IndyInterface.fromCache(IndyInterface.java:321)
	at nextflow.extension.CollectFileOp.emitItems(CollectFileOp.groovy:187)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:343)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:328)
	at groovy.lang.MetaClassImpl.doInvokeMethod(MetaClassImpl.java:1333)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1088)
	at groovy.lang.MetaClassImpl.invokeMethodClosure(MetaClassImpl.java:1017)
	at groovy.lang.MetaClassImpl.doInvokeMethod(MetaClassImpl.java:1207)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1088)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1007)
	at groovy.lang.Closure.call(Closure.java:433)
	at groovy.lang.Closure.call(Closure.java:422)
	at org.codehaus.groovy.vmplugin.v8.IndyInterface.fromCache(IndyInterface.java:321)
	at nextflow.extension.DataflowHelper$3.afterStop(DataflowHelper.groovy:257)
	at groovyx.gpars.dataflow.operator.DataflowProcessor.fireAfterStop(DataflowProcessor.java:324)
	at groovyx.gpars.dataflow.operator.DataflowProcessorActor.afterStop(DataflowProcessorActor.java:59)
	...
```

This might be addressable using existing Nextflow configuration settings `google.httpReadTimeout` and `google.httpConnectTimeout` so I can retry but let me know if you have any further insights. Log is attached.
[KwJ0dUQ6dCkN7.log](https://github.com/user-attachments/files/18646343/KwJ0dUQ6dCkN7.log)

**Comment by pditommaso** [comment]
It would be preferable to keep the focus on the Batch API (and address this one separately). 

@ejseqera is the batch execution performing better is your tests or you could not assess due the above error?

File: modules/nextflow/src/main/groovy/nextflow/processor/TaskArrayCollector.groovy [hunk 5]
```diff
@@ -91,11 +91,13 @@ class TaskArrayCollector {
             // submit task directly if the collector is closed
             // or if the task is retried (since it might have dynamic resources)
             if( closed || task.config.getAttempt() > 1 ) {
+                task.isChild = false
```
Comment (pditommaso) [inline modules/nextflow/src/main/groovy/nextflow/processor/TaskArrayCollector.groovy:94]: This is implicit because `boolean` is `false` by default

File: modules/nextflow/src/main/groovy/nextflow/processor/TaskArrayCollector.groovy [hunk 6]
```diff
@@ -116,7 +118,9 @@ class TaskArrayCollector {
         sync.lock()
         try {
             if( array.size() == 1 ) {
-                executor.submit(array.first())
+                final task = array.first()
+                task.isChild = false
```
Comment (pditommaso) [inline modules/nextflow/src/main/groovy/nextflow/processor/TaskArrayCollector.groovy:122]: not needed as above

File: modules/nextflow/src/main/groovy/nextflow/processor/TaskArrayCollector.groovy (hunk 5 above)
Comment (jorgee) [reply modules/nextflow/src/main/groovy/nextflow/processor/TaskArrayCollector.groovy:94]: I have explicitly set to false to ensure it is not treat as an array when task is retried.

**Comment by ejseqera** [comment]
I ran a large-scale test (57gB7BLSjrLuXj, ~1200 tasks) and it looks like the original `NotFoundException` issue appears to be resolved with this implementation. 

During testing, I did encounter some other exceptions, but these are separate from the `NotFoundException` issue this PR addresses:

1. `DeadlineExceededException` from the Batch API after ~1200 successful tasks
- The error occurred when the gRPC call to `batch.googleapis.com` exceeded its 60-second deadline. This happened despite having configured a retry policy:
```
      retryPolicy {
         maxDelay = '60s'
         maxAttempts = 15
         jitter = 0.5
         delay = '5s'
      }
```
   - This can be mitigated with alternative retry policy configuration
   - Not related to the original task status resolution problem

3. `RejectedExecutionException` during pipeline shutdown
   - Related to thread pool management during termination
   - Again, separate from the task status handling

These new findings could be addressed separately through configuration or future PRs if needed, but I don't think they impact the effectiveness of this PR's solution for the `NotFoundException` issue. What do you think? @pditommaso @jorgee 

Logs for two separate runs in different regions attached.
[57gB7BLSjrLuXj.log](https://github.com/user-attachments/files/18734206/57gB7BLSjrLuXj.log)
[20EJcGulEDB2GF.log](https://github.com/user-attachments/files/18734207/20EJcGulEDB2GF.log)

File: modules/nextflow/src/main/groovy/nextflow/exception/ProcessStageException.groovy [hunk 7]
```diff
@@ -24,5 +24,5 @@ import groovy.transform.InheritConstructors
  * @author Paolo Di Tommaso <paolo.ditommaso@gmail.com>
  */
 @InheritConstructors
-class ProcessStageException extends ProcessException implements ShowOnlyExceptionMessage {
+class ProcessStageException extends ProcessUnrecoverableException implements ShowOnlyExceptionMessage {
```
Comment (pditommaso) [inline modules/nextflow/src/main/groovy/nextflow/exception/ProcessStageException.groovy:27]: What's the rationale to change this to "unrecoverable" ?

Comment (jorgee) [reply modules/nextflow/src/main/groovy/nextflow/exception/ProcessStageException.groovy:27]: It is related to #5727. When there is a problem with data staging, nextflow retries the execution but hash is null. ProcessStageException is changed to unrecoverable to avoid this task retries.

Comment (jorgee) [reply modules/nextflow/src/main/groovy/nextflow/exception/ProcessStageException.groovy:30]: Done
